tests/dt.go

```diff
@@ -0,0 +1,234 @@
+	DownPeers       []*pdpb.PeerStats `json:"down_peers,omitempty"`
+	PendingPeers    []*metapb.Peer    `json:"pending_peers,omitempty"`
+	WrittenBytes    uint64            `json:"written_bytes,omitempty"`
+	ReadBytes       uint64            `json:"read_bytes,omitempty"`
+	ApproximateSize int64             `json:"approximate_size,omitempty"`
+	ApproximateKeys int64             `json:"approximate_keys,omitempty"`
+}
+
+// RegionsInfo contains some regions with the detailed region info.
+type RegionsInfo struct {
+	Count   int           `json:"count"`
+	Regions []*RegionInfo `json:"regions"`
+}
+
+func (oa *operatorActions) LabelNodes() error {
+	nodes, err := oa.kubeCli.CoreV1().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for i, node := range nodes.Items {
+		index := i % RackNum
+		node.Labels[RackLabel] = fmt.Sprintf("rack%d", index)
+		oa.kubeCli.CoreV1().Nodes().Update(&node)
+	}
+	return nil
+}
+
+func (oa *operatorActions) LabelNodesOrDie() {
+	err := oa.LabelNodes()
+	if err != nil {
+		slack.NotifyAndPanic(err)
+	}
+}
+
+func (oa *operatorActions) CheckDisasterTolerance(cluster *TidbClusterConfig) error {
+	nodeMap := map[string]corev1.Node{}
+	nodes, err := oa.kubeCli.CoreV1().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, node := range nodes.Items {
+		nodeMap[node.Name] = node
+	}
+
+	pds, err := oa.kubeCli.CoreV1().Pods(cluster.Namespace).List(
+		metav1.ListOptions{LabelSelector: labels.SelectorFromSet(
+			label.New().Instance(cluster.ClusterName).PD().Labels(),
+		).String()})
+	if err != nil {
+		return err
+	}
+	err = oa.checkDisasterTolerance(pds.Items, nodeMap)
+	if err != nil {
+		return err
+	}
+
+	tikvs, err := oa.kubeCli.CoreV1().Pods(cluster.Namespace).List(
+		metav1.ListOptions{LabelSelector: labels.SelectorFromSet(
+			label.New().Instance(cluster.ClusterName).TiKV().Labels(),
+		).String()})
+	if err != nil {
+		return err
+	}
+	err = oa.checkDisasterTolerance(tikvs.Items, nodeMap)
+	if err != nil {
+		return err
+	}
+
+	tidbs, err := oa.kubeCli.CoreV1().Pods(cluster.Namespace).List(
+		metav1.ListOptions{LabelSelector: labels.SelectorFromSet(
+			label.New().Instance(cluster.ClusterName).TiDB().Labels(),
+		).String()})
+	if err != nil {
+		return err
+	}
+	return oa.checkDisasterTolerance(tidbs.Items, nodeMap)
+}
+
+func (oa *operatorActions) checkDisasterTolerance(allPods []corev1.Pod, nodeMap map[string]corev1.Node) error {
+	rackPods := map[string][]corev1.Pod{}
+	for _, pod := range allPods {
+		if node, exist := nodeMap[pod.Spec.NodeName]; exist {
+			pods, exist := rackPods[node.Labels[RackLabel]]
+			if !exist {
+				pods = []corev1.Pod{}
+			}
+			pods = append(pods, pod)
+			rackPods[node.Labels[RackLabel]] = pods
+		}
+	}
+
+	podNum := len(allPods)
+	maxPodsOneRack := podNum / RackNum
+	mod := podNum % RackNum
+	if mod > 0 {
+		maxPodsOneRack = maxPodsOneRack + 1
+	}
+
+	for rack, pods := range rackPods {
+		if len(pods) > maxPodsOneRack {
```

Comment:
I think we should also check minimum pods.

Consider 4 pod with 3 rack, max pod is 2, [2, 2, 0] fit the criteria but not the best arrangement to be fault-tolerant. Add the minimum check `>=1` will help.

Reply:
nice catch! it's a bug